### PR TITLE
Functionality to add files as a resource from configuration management rest API

### PR DESCRIPTION
**Pull request by Buddhimah**
Proposed changes in this pull request
Fix: wso2/product-is#6146

This will add the functionality to add files as a resource from configuration management rest API

This pull request adds file management feature in wso2/product-is#4303

### CODE REVIEW

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java [hunk 1]
```diff
@@ -706,4 +736,189 @@ private void checkFeatureStatus() throws ConfigurationManagementException {
                     ErrorMessages.ERROR_CODE_FEATURE_NOT_ENABLED.getCode());
         }
     }
+
+    /**
+     * {@inheritDoc}
```
[the remark this answers is not in the document]
Comment (Buddhimah) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java]: removed

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java [hunk 2]
```diff
@@ -706,4 +736,189 @@ private void checkFeatureStatus() throws ConfigurationManagementException {
...
+     */
+    @Override
+    public ResourceFile addFile(String resourceTypeName, String resourceName, String fileName, InputStream fileStream)
+            throws ConfigurationManagementException {
+
+        validateFileAddRequest(resourceTypeName, resourceName, fileName, fileStream);
+        String resourceId = getResourceId(resourceTypeName, resourceName);
+        String fileId = generateUniqueID();
+        if (log.isDebugEnabled()) {
+            log.debug("File id generated: " + fileId);
```
[the remark this answers is not in the document]
Comment (Buddhimah) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java]: done

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java [hunk 3]
```diff
@@ -706,4 +736,189 @@ private void checkFeatureStatus() throws ConfigurationManagementException {
...
+        }
+        getConfigurationDAO().addFile(fileId, resourceId, fileName, fileStream);
+        if (log.isDebugEnabled()) {
+            log.debug("File: " + fileId + " successfully added.");
```
[the remark this answers is not in the document]
Comment (Buddhimah) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java]: done

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java [hunk 4]
```diff
@@ -706,4 +736,189 @@ private void checkFeatureStatus() throws ConfigurationManagementException {
...
+        }
+        return new ResourceFile(fileId, getFilePath(fileId, resourceTypeName, resourceName).toString(), fileName);
```
[the remark this answers is not in the document]
Comment (Buddhimah) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java]: done

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java [hunk 5]
```diff
@@ -706,4 +736,189 @@ private void checkFeatureStatus() throws ConfigurationManagementException {
...
+        validateFilesDeleteRequest(resourceTypeName, resourceName);
+        String resourceId = getResourceId(resourceTypeName, resourceName);
+        getConfigurationDAO().deleteFiles(resourceId);
+        if (log.isDebugEnabled()) {
+            log.debug("All the files were deleted in the resource: " + resourceName + ".");
+        }
+    }
+
+    /**
+     * {@inheritDoc}
```
[the remark this answers is not in the document]
Comment (Buddhimah) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManagerImpl.java]: done

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManager.java [hunk 6]
```diff
@@ -188,4 +192,82 @@ Attribute getAttribute(String resourceTypeName, String resourceName, String attr
...
+     * @return 200 ok.
+     * @throws ConfigurationManagementException Resource management exception.
+     */
+    void deleteFiles(String resourceTypeName, String resourceName)
+            throws ConfigurationManagementException;
+
+    /**
+     * This API is used to get the given file.
+     *
+     *
```
Comment (tharindu-bandara) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManager.java]: Redundant new line. Please check other places as well.

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java [hunk 7]
```diff
@@ -960,4 +992,226 @@ private Timestamp getCreatedTimeInResponse(Resource resource, String resourceTyp
...
+    }
+
+    @Override
+    public InputStream getFileById(String resourceType, String resourceName, String fileId) throws ConfigurationManagementException {
+
+        JdbcTemplate jdbcTemplate = JdbcUtils.getNewTemplate();
+        try {
+            return jdbcTemplate.fetchSingleRecord(SQLConstants.GET_FILE_BY_ID_SQL,
+                    (resultSet, rowNumber) -> {
+                        Blob fileBlob = resultSet.getBlob("VALUE");
```
Comment (tharindu-bandara) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java]: let's use a constant for "VALUE"

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java [hunk 8]
```diff
@@ -960,4 +992,226 @@ private Timestamp getCreatedTimeInResponse(Resource resource, String resourceTyp
...
+     * @return A list of {@link Attribute} for the given resource.
+     */
+    private List<Attribute> getAttributesByResourceId(String resourceId)
+            throws ConfigurationManagementServerException {
+
+        JdbcTemplate jdbcTemplate = JdbcUtils.getNewTemplate();
+        try {
+            return jdbcTemplate.executeQuery(GET_ATTRIBUTES_BY_RESOURCE_ID_SQL,
+                    ((resultSet, rowNumber) -> {
+                        String attributeId = resultSet.getString("ID");
```
Comment (tharindu-bandara) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java]: Let's use constants

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationRawDataCollector.java [hunk 9]
```diff
@@ -288,5 +295,14 @@ public ConfigurationRawDataCollectorBuilder setFileId(String fileId) {
             this.fileId = fileId;
             return this;
         }
+
+        public String getFileName() {
+            return fileName;
+        }
+
+        public ConfigurationRawDataCollectorBuilder setFileName(String fileName) {
+            this.fileName = fileName;
```
Comment (tharindu-bandara) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationRawDataCollector.java]: Let's add a new line after the method

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/model/ResourceFile.java [hunk 10]
```diff
@@ -21,24 +21,69 @@
...
+
+        this.id = id;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public void setName(String name) {
+        this.name = name;
```
Comment (tharindu-bandara) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/model/ResourceFile.java]: let's add a new line here. check other places

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java (hunk 7 above)
Comment (Buddhimah) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java]: done

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManager.java (hunk 6 above)
Comment (Buddhimah) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/ConfigurationManager.java]: done

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/constant/ConfigurationConstants.java [hunk 11]
```diff
@@ -102,7 +112,16 @@
         ERROR_CODE_REPLACE_RESOURCE("CONFIGM_00034", "Error while replacing the resource : %s."),
         ERROR_CODE_REPLACE_ATTRIBUTE("CONFIGM_00035", "Error while replacing the attribute: %s."),
         ERROR_CODE_FEATURE_NOT_ENABLED("CONFIGM_00036", "Configuration management feature is not enabled."),
-        ;
+        ERROR_CODE_INSERT_FILE("CONFIGM_00037", "Error while adding the file: %s."),
+        ERROR_CODE_GET_FILE("CONFIGM_00038", "Error while retrieving the file: %s."),
+        ERROR_CODE_GET_FILES("CONFIGM_00039", "Error while retrieving files the resource: %s."),
```
Comment (denuwanthi) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/constant/ConfigurationConstants.java]: "Error while retrieving files **for** the resource: %s."?

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/util/ConfigurationUtils.java [hunk 12]
```diff
@@ -130,12 +147,54 @@ public static boolean useCreatedTimeField() {
...
+     * Creates the file endpoint using the parameters.
+     *
+     * @param fileId file id.
+     * @param resourceName resource name.
+     * @param resourceType resource type name.
+     * @return tenant domain of the request is being served.
+     */
+    public static String getFilePath(String fileId, String resourceType, String resourceName) {
+
+        return RESOURCE_PATH + '/' + resourceType + '/' + resourceName + '/' + FILE + '/' + fileId;
```
Comment (denuwanthi) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/util/ConfigurationUtils.java]: Shall we use constant for '/' ?

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/constant/ConfigurationConstants.java [hunk 13]
```diff
@@ -102,7 +112,16 @@
...
+        ERROR_CODE_DELETE_FILES("CONFIGM_00040", "Error while deleting files for the resource id: %s."),
+        ERROR_CODE_DELETE_FILE("CONFIGM_00041", "Error while deleting the file with id: %s."),
+        ERROR_CODE_GET_FILES_BY_TYPE("CONFIGM_00042", "Error while retrieving files the resource type: %s."),
```
Comment (sachiniWettasinghe) [inline components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/constant/ConfigurationConstants.java]: This should be "Error while retrieving files **for** the resource type: %s."

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/util/ConfigurationUtils.java (hunk 12 above)
Comment (Buddhimah) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/util/ConfigurationUtils.java]: FIxed https://github.com/wso2/carbon-identity-framework/pull/2690

File: components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/constant/ConfigurationConstants.java (hunk 13 above)
Comment (Buddhimah) [reply components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/constant/ConfigurationConstants.java]: Fixed https://github.com/wso2/carbon-identity-framework/pull/2690

**Pull request merged**